### PR TITLE
Wave color nonsense

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: #5652
Depends on #5828

QA:
- [x] Changing of wave track color with the track's menu
- [x] Mixing-and-rendering keeps color setting of the first mixed track;  then splitting to mono channels keeps same in both
- [ ] Recording or generating into a track uses the wave track color
- [x] Dragging clips from track to track preserves clip color, even if one track then has multiple colors
- [x] Mix-and-render names the new track as before (either "Mix" + number or keeping name of the sole track)
- [x] "Get Info" macro command for clips reports correct colors
- [x] "Set Clip" macro command, for colors
- [x] "Set Track Visuals" macro command, for colors
- [x] Cut/Copy then Paste of tracks (in or cross-project) preserves colors
- [x] Pasting or generating sound into an existing clip does not change its color
- [x] When a clip lengthens for sync lock with another track, its color does not change
- [x] Splitting a clip colors both as the original
- [x] Undo and Redo of all of the above
- [x] Waveform, Spectrogram panning and zooming are not slower
- [x] Persistency of wave track color settings, and of clip color
- [x] Opening in 3.4, a project that was saved in 3.5, works with correct colors; and oppositely

*(short description of the changes and the motivation to make the changes)*

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: libraries/lib-wave-track/WaveTrack.h
Comment (vsverchinsky) [inline libraries/lib-wave-track/WaveTrack.h:1370]: `deserialization` flag feels like extra parameter

File: src/tracks/playabletrack/wavetrack/ui/WaveformView.cpp
Comment (vsverchinsky) [inline src/tracks/playabletrack/wavetrack/ui/WaveformView.cpp:1204]: My guess is that there will be an extra call to `SetColorIndex` during deserialization: one from `OnNewClip`, and second in `HandleXMLAttribute`, right?

File: libraries/lib-wave-track/WaveClip.h
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveClip.h:91]: As linked in CRTPBase.h, it's recommended here:

https://learn.microsoft.com/en-us/cpp/cpp/general-rules-and-limitations?view=msvc-170

I use the same in `extract-lib-note-track` which has passed code review but is not yet merged into master.

Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveClip.h:91]: See the bottom of that page

File: libraries/lib-wave-track/WaveClip.cpp
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveClip.cpp:36]: I see the same in the logs for the build of `extract-lib-note-track`.  I think I should instead give an inline '{}' body to the dtor in the class definition and let the linker de-duplicate it.

File: libraries/lib-wave-track/WaveTrack.h
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.h:1365]: How would I create the observers and the subscriptions if WaveTrack is the publisher?

Each new WaveTrack must inform its own listener, which has its own separate state (the wave track color), of the event of addition of clips.

So the listener must be created too with each WaveTrack -- and that's where `ClientData` is useful, if also I use `BuildAll` instead of creation on demand.  It allows hooking of a constructor of a class.

And I need `BuildAll` so that the new attached object is present from the beginning, before any clips are created or deserialized.

But you are right that I could then also use Observer instead of this class.

There is the problem of getting deep-copy right, but that is what `Reparent` is for.

File: libraries/lib-wave-track/WaveTrack.h
Comment (Paul-Licameli) [reply libraries/lib-wave-track/WaveTrack.h:1370]: I'm not sure of the complete set of messages that will be useful in future.  I could easily use Observer and an enumeration now for creation, deserialization, deletion, ...

File: src/tracks/playabletrack/wavetrack/ui/WaveformView.cpp
Comment (Paul-Licameli) [reply src/tracks/playabletrack/wavetrack/ui/WaveformView.cpp:1204]: Yes, does it matter much?

It would matter in case a file serialized wave color without clip color, but then the clip colors would still be initialized.

I don't think there are any such legacy files, though.

**Comment by Paul-Licameli** [comment]
Rebased to recent `audacity/wave-clip-refactoring` and resolved an easy conflict

File: libraries/lib-wave-track/WaveClip.h
Comment (crsib) [reply libraries/lib-wave-track/WaveClip.h:91]: I think we should revise the usage of CRTPBase in a separate PR. I have opened a task for that #5704

**Comment by chinakov** [comment]
If a User has a non-default color for a Stereo track, and then:
1. Swaps Stereo Channel - the audio track reverts back to default color

**Comment by chinakov** [comment]
If a User has a **non-default** color for a Stereo track, and then:
1. Swaps Stereo Channels - the audio track reverts back to default color (blue)
2. Splits Stereo Track - the last audio track reverts back to default color (blue)
3. Splts track to Mono - the last audio track reverts back to default color (blue)

This behaviour has been changed from the last release version. 

If a User has non-default colors for 2 Mono tracks, the behaviours for joining and mixing and rendering are not changed from the last release version.

**Comment by chinakov** [comment]
> Recording or generating into a track uses the wave track color
Generating indeed does use the wave track color, but recording into said track still preserves the default track color (blue). 
Pictured: Recording preserve default blue color, generated audio preserves custom color defined by the User.
image

**Comment by chinakov** [comment]
Another issue: if a User opens a project with differently coloured clips within one track, one of the channels will be reverted back to default color. 

How it was saved in your Branch Build:
image


How it was opened in Current Release Version:
image

**Comment by Paul-Licameli** [comment]
> If a User has a **non-default** color for a Stereo track, and then:
> 
>     1. Swaps Stereo Channels - the audio track reverts back to default color (blue)
> 
>     2. Splits Stereo Track - the last audio track reverts back to default color (blue)
> 
>     3. Splts track to Mono - the last audio track reverts back to default color (blue)
> 
> 
> This behaviour has been changed from the last release version.
> 
> If a User has non-default colors for 2 Mono tracks, the behaviours for joining and mixing and rendering are not changed from the last release version.

Considering this problem, I now realize that this issue must depend first on fixing #5530

File: src/tracks/playabletrack/wavetrack/ui/WaveformCache.h
Comment (crsib) [inline src/tracks/playabletrack/wavetrack/ui/WaveformCache.h:41]: Serializable cache is not a usual thing, TBH

File: libraries/lib-wave-track/WaveClip.h
Comment (crsib) [inline libraries/lib-wave-track/WaveClip.h:100]: I think this violates SRP and in a bad way. I missed it somehow, during the first pass of the review, but it caught my eye when I saw color index serialization in the wave cache.

**Review by crsib — COMMENTED** [review]
I am really confused by this PR, I will wait before the parent branch is merged and this PR is rebased before giving it a new round of review.

Reviewing it commit-by-commit makes it hard to understand if there is a simple way to improve it and I am seriously concerned by some of the decisions made in it. 

I cannot approve it as it is.

**Comment by Paul-Licameli** [comment]
As you understood, "Cache" is now a misnomer and the next commit renames them as "Attachments" to WaveClip.  The are attached data on the clip getting certain notifications of events in the clip's lifetime.

Some attachment on WaveClip removes the irrelevant mention of color from lib-wave-track which shouldn't be there.

I did reuse existing WaveClipWaveformCache to define the override of WriteXMLAttrubutes.  If you don't like it, it's easily done to make another class just to hold the color with its own registered factory.  I didn't see that as necessary.

I don't like it that clips are now mentioning stretches and pitch shifts (also serialized data) and would put them in WaveClip attachments too.

**Comment by Paul-Licameli** [comment]
Force pushed with changes that separate the attachments for colors, from the other attachments that really perform only caching.

**Comment by crsib** [comment]
Looks like you forgot to push.

**Comment by Paul-Licameli** [comment]
Rebased without change to updated wave-clip-refactoring branch

File: libraries/lib-wave-track/WaveClip.h
Comment (crsib) [reply libraries/lib-wave-track/WaveClip.h:100]: I have a strong opinion that the fix is worse than the problem. I understand that it is pointless to argue.

**Review by crsib — APPROVED** [review]
This is an incredibly complex PR addressing a minor problem. I don't like some decisions made here, but I don't want to block your branch either.

**Comment by Paul-Licameli** [comment]
Rebased without changes

**Comment by Paul-Licameli** [comment]
One new commit for review

**Comment by Paul-Licameli** [comment]
Rebased onto updated dependency branch

**Review by crsib — APPROVED** [review]

**Comment by Paul-Licameli** [comment]
Rather than simply rewriting SetWaveColorIndex as interval-major, better to remove completely from lib-wave-track, any mention of extraneous display related things.

Some surprisingly roundabout solutions are needed with attached objects on WaveTrack and WaveClip, but they may prove useful in future to attach other extra information to tracks and clips non-intrusively.

**Pull request merged**
